File: 00xx/000x/0008-StringToIntegerAtoi/Solution.java
/*  Solution: Character-by-character
    Time complexity:    O(n^3)
    Space complexity:   O(1)
 */

class Solution {    
    public int myAtoi(String str) {
        boolean startedInt = false;
        boolean isNegative = false;
        int val = 0;
        for (int i = 0; i < str.length(); i++) {
            //  Case for digits 0-9
            if (str.charAt(i) == '0' ||
                str.charAt(i) == '1' ||
                str.charAt(i) == '2' ||
                str.charAt(i) == '3' ||
                str.charAt(i) == '4' ||
                str.charAt(i) == '5' ||
                str.charAt(i) == '6' ||
                str.charAt(i) == '7' ||
                str.charAt(i) == '8' ||
                str.charAt(i) == '9'
            ) {
                startedInt = true;
                int digit = 0;
                switch (str.charAt(i)) {
                    case '0':
                        digit = 0;
                        break;
                    case '1':
                        digit = 1;
                        break;
                    case '2':
                        digit = 2;
                        break;
                    case '3':
                        digit = 3;
                        break;
                    case '4':
                        digit = 4;
                        break;
                    case '5':
                        digit = 5;
                        break;
                    case '6':
                        digit = 6;
                        break;
                    case '7':
                        digit = 7;
                        break;
                    case '8':
                        digit = 8;
                        break;
                    case '9':
                        digit = 9;
                        break;
                }
                //  Check if return value is about to overflow
                int bound = (isNegative ? Integer.MIN_VALUE % 10 * -1 :
                             Integer.MAX_VALUE % 10);
                if (val > Integer.MAX_VALUE / 10 || (
                    val == Integer.MAX_VALUE / 10 &&
                    digit > bound
                )) {
                    return (
                        isNegative ?
                        Integer.MIN_VALUE :
                        Integer.MAX_VALUE
                    );
                }
                else {
                    val = val * 10 + digit;
                }
            }
            //  Handle +/- and toggle isNegative flag accordingly
            else if (str.charAt(i) == '+' || str.charAt(i) == '-') {
                if (startedInt) {
                    return isNegative ? -val : val;
                }
                else {
                    isNegative = (str.charAt(i) == '-') ? true : false;
                    startedInt = true;
                }
            }
            //  Handle pre-integer whitespaces
            else if (str.charAt(i) == ' ') {
                if (startedInt) {
                    return isNegative ? -val : val;
                }
                else {
                    continue;
                }
            }
            else {
                return isNegative ? -val : val;
            }
        }
        return isNegative ? -val : val;
    }
}

File: 00xx/000x/0008-StringToIntegerAtoi/Solution.go
/*  Solution: Lookup Hash Map for each character of input
    Time Complexity:    O(n+1)  (n is length of str, +1 for negative check)
    Space Complexity:   O(1)  (constant memory)
 */

func myAtoi(str string) int {
  handleSpaces := true
  isNeg := false
  val := 0
  const INT_MAX = int(^uint32(0) >> 1)
  table := map[rune]int{
    '0': 0,
    '1': 1,
    '2': 2,
    '3': 3,
    '4': 4,
    '5': 5,
    '6': 6,
    '7': 7,
    '8': 8,
    '9': 9,
  }
  for _, char := range str {
    if char == ' ' {
      if handleSpaces == false {
        break
      }
    } else if char == '-' || char == '+' {
      if handleSpaces == false {
        break
      }
      if char == '-' {
        isNeg = true
      }
      handleSpaces = false
    } else {
      if digit, ok := table[char]; ok {
        if isNeg == false && (
          val > INT_MAX / 10 || (
            val == INT_MAX / 10 &&
            digit > INT_MAX % 10)) {
          return INT_MAX
        } else if isNeg == true && (
          val > INT_MAX / 10 || (
            val == INT_MAX / 10 &&
            digit > INT_MAX % 10 + 1)) {
          return -INT_MAX - 1
        } else {
          val = val * 10 + digit
        }
      } else {
        break
      }
      handleSpaces = false
    }
  }
  if isNeg == true {
    val *= -1
  }
  return val
}
